### PR TITLE
#3127 storage symlink feature

### PR DESCRIPTION
**Pull request by cboudereau**
The impl is done
I have manually tested on mono + windows the behavior (paket local, restore, install and update) with and without the configuration / globally and on a specific package.

On a little project 5s before and 270ms after.

When you restore dotnet core packages and the nuget cache is full the restore phase is less a second instead of 20s.

I am building integration test.

### CODE REVIEW

**Comment by cboudereau** [comment]
@forki I don't know why there was an error on the Travis ci build with mono. Could you help me ? The metaproj error is a little bit weird for me. Thanks!

File: Paket.preview3.sln [hunk 1]
```diff
@@ -1,71 +1,86 @@
-
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
 VisualStudioVersion = 15.0.26730.0
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "Paket.Core", "src/Paket.Core.preview3/Paket.Core.fsproj", "{779DA2DD-CEA0-4EC4-9DBD-2CF29C2269EA}"
+Project("{6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705}") = "Paket.Core", "src\Paket.Core.preview3\Paket.Core.fsproj", "{779DA2DD-CEA0-4EC4-9DBD-2CF29C2269EA}"
```
Comment (matthid) [inline Paket.preview3.sln]: change `\` back to `/` and travis will likely work again...

Comment (cboudereau) [reply Paket.preview3.sln]: Oh thanks a lot! @matthid

File: src/Paket.Core/Dependencies/NuGet.fs [hunk 2]
```diff
@@ -806,21 +806,36 @@ let private downloadAndExtractPackage(alternativeProjectRoot, root, isLocalOverr
                 | exn -> raise (Exception(sprintf "Could not download %O %O from %s." packageName version !downloadUrl, exn)) }
 
     async {
+        configResolved.Path |> Option.iter SymlinkUtils.delete
+
         do! download true 0
-        if not isLocalOverride then
+
+        match isLocalOverride, configResolved with
+        | true, ResolvedPackagesFolder.NoPackagesFolder -> return failwithf "paket.local in combination with storage:none is not supported"
```
Comment (matthid) [inline src/Paket.Core/Dependencies/NuGet.fs]: Maybe we add to that message "use storage: symlink instead" ?

File: src/Paket.Core/Installation/RestoreProcess.fs [hunk 3]
```diff
@@ -111,7 +111,7 @@ let ExtractPackage(alternativeProjectRoot, root, groupName, sources, caches, for
             | LocalNuGet(path,_) as source ->
                 return! extractPackage caches package alternativeProjectRoot root localOverride source groupName v includeVersionInPath downloadLicense force
         }
-
+
```
Comment (matthid) [inline src/Paket.Core/Installation/RestoreProcess.fs]: maybe remove that whitespace

File: src/Paket.Core/PaketConfigFiles/RuntimeGraph.fs [hunk 4]
```diff
@@ -203,14 +203,13 @@ module RuntimeGraph =
 
     open System.IO
     /// Downloads the given package into the nuget cache and read its runtime.json.
-    let getRuntimeGraphFromNugetCache root groupName (package:ResolvedPackage) =
-        let config = PackagesFolderGroupConfig.NoPackagesFolder
+    let getRuntimeGraphFromNugetCache root configO groupName (package:ResolvedPackage) =
```
Comment (matthid) [inline src/Paket.Core/PaketConfigFiles/RuntimeGraph.fs]: `configO`? Maybe just `config`?

**Comment by matthid** [comment]
One could argue about starting external processes, but we already do that on other places as well.

File: src/Paket.Core/Installation/RestoreProcess.fs (hunk 3 above)
Comment (cboudereau) [reply src/Paket.Core/Installation/RestoreProcess.fs]: done

File: src/Paket.Core/PaketConfigFiles/RuntimeGraph.fs (hunk 4 above)
Comment (cboudereau) [reply src/Paket.Core/PaketConfigFiles/RuntimeGraph.fs]: done

**Comment by cboudereau** [comment]
Yeah about external process I agree but dotnet core API are not yet ready due to case clash between claimed and FileInfo/DirectoryInfo (https://github.com/dotnet/corefx/issues/25569). So I reuse the external tools to preserve OS dependent behavior.

File: src/Paket.Core/Dependencies/NuGet.fs (hunk 2 above)
Comment (cboudereau) [reply src/Paket.Core/Dependencies/NuGet.fs]: done

**Comment by cboudereau** [comment]
@matthid I have fixed the last failing test. There is some well known tests failing on AppVeyor (as I understand according to previous merged). Is it ok for you ? Thanks in advance your help! I appreciated it

**Comment by cboudereau** [comment]
Hello @matthid, do you think that the checks are done (even if there is well known failing tests ? I am currently investigating why those tests failed)

**Comment by cboudereau** [comment]
Here is the diagnostic : 

" Lowercase package names in package cache: old csproj, packages folder enabled" is failing due to FSharp.Core binding redirects (4.4.1 -> 4.4.3) of the FakeLib.dll logger. By disabling logger the test works fine (Fake.MSBuildHelper.MSBuildLoggers <- [] //There is a fsharp.core binding redirect issue on the FakeLib.dll logger)

"#2335 should install deps from different groups when using conditions" : this one failed due to new framework versions.

"#2294 pin NETStandard.Library = 1.6.0 Strategy Workaround" : paket timeout after 7mn
"2294 Cannot pin NETStandard.Library = 1.6.0" : paket timeout after 7mn

I am running with a timeout infinite the #2294 but test are still running for 20mn.

If I compare the build tests between mine and the last merged, tests are on the same status.
https://ci.appveyor.com/project/paket/paket/build/0.0.1.347

I am adding a new commit to fix 2 tests.

**Comment by cboudereau** [comment]
@matthid by fixing those 2 tests (not flaky), all CI are done. @forki could you check if all is ok please ?

Thank a lots for your help!

**Comment by forki** [comment]
I wonder if we can make that the default eventually?

**Comment by thinkbeforecoding** [comment]
Probably. Let's make it opt-in for now, and make it default later when we are sure there are no problematic edge cases.

**Comment by forki** [comment]
I get:

image

**Comment by cboudereau** [comment]
Which OS/Environment ? (as I understand it is a permission issue). We are currently running it on windows without highest privilege. Same on bash ubuntu for windows.

**Comment by forki** [comment]
running on windows in non-admin terminal

**Comment by cboudereau** [comment]
Have you checked policies around mklink : Local Policies\User Rights Assignment\Create symbolic links ??

**Comment by matthid** [comment]
So no default for now ;)

**Comment by matthid** [comment]
But it is probably fine as it is. I don't think it will get any better than this.

**Comment by thinkbeforecoding** [comment]
Are you sure there's not some process locking the dll ?

**Comment by thinkbeforecoding** [comment]
We reproduce the problem on our CI server on a build.. checking what's happening.

**Comment by cboudereau** [comment]
By configuring the local policy it works fine

**Comment by forki** [comment]
@cboudereau can you please send a PR to the docs with details? also a error message that points to the docs would be nice

**Comment by cboudereau** [comment]
Sure

**Comment by cboudereau** [comment]
done #3134

**Pull request merged**